### PR TITLE
Add support for containers.conf

### PR DESCRIPTION
**Pull request by rhatdan**
vendor in c/common config pkg for containers.conf

Signed-off-by: Qi Wang <qiwan@redhat.com>
Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
Replaces https://github.com/containers/libpod/pull/4569

**Comment by giuseppe** [comment]
these two PRs should fix the issue you are seeing:

https://github.com/containers/libpod/pull/4911
https://github.com/containers/common/pull/46

**Comment by raukadah** [comment]
From RHEL-8 rpm package build job, we are seeing from log link: http://logs.rdoproject.org/98/4698/7ff850dfccafcf1ed1bd7c3decdfb5e3e627b9fb/github-check/podman-package-rhel-8/101e05b/job-output.txt
```
2020-01-23 00:06:56.465249 | test | GO111MODULE=on go build -mod=vendor  -gcflags 'all=-trimpath=/root/rpmbuild/BUILD/libpod-7ff850df' -asmflags 'all=-trimpath=/root/rpmbuild/BUILD/libpod-7ff850df' -ldflags '-X github.com/containers/libpod/libpod/define.gitCommit=8fe939424f8b2cf86d38c576b2a2e2d5fe7f1309-dirty -X github.com/containers/libpod/libpod/define.buildInfo=1579756016 -X github.com/containers/libpod/libpod/config._installPrefix=/usr/local -X github.com/containers/libpod/libpod/config._etcDir=/etc -extldflags ""' -tags "varlink selinux seccomp exclude_graphdriver_btrfs btrfs_noversion  exclude_graphdriver_devicemapper" -o bin/podman github.com/containers/libpod/cmd/podman
2020-01-23 00:07:32.430188 | test | # github.com/containers/libpod/pkg/api/handlers
2020-01-23 00:07:32.430499 | test | pkg/api/handlers/images_build.go:185:3: unknown field 'AddCapabilities' in struct literal of type imagebuildah.BuildOptions
2020-01-23 00:07:32.430521 | test | pkg/api/handlers/images_build.go:186:3: unknown field 'DropCapabilities' in struct literal of type imagebuildah.BuildOptions
2020-01-23 00:07:33.218508 | test | make[1]: *** [Makefile:174: bin/podman] Error 2
2020-01-23 00:07:33.219326 | test | make[1]: Leaving directory '/root/rpmbuild/BUILD/libpod-7ff850df'
2020-01-23 00:07:33.222341 | test | error: Bad exit status from /var/tmp/rpm-tmp.kIedqm (%build)
``

**Comment by rhatdan** [comment]
@raukadah That should be fixed now.

**Comment by vrothberg** [comment]
Once merged, we may close https://github.com/containers/libpod/issues/3587

**Comment by vrothberg** [comment]
... and https://github.com/containers/libpod/issues/3464

**Comment by rhatdan** [comment]
Fixes:https://github.com/containers/libpod/issues/4084

**Comment by TomSweeneyRedHat** [comment]
@rhatdan looks like gating is complaining of an incomplete man page change.  Should this still be a WIP?

**Comment by rhatdan** [comment]
@TomSweeneyRedHat Just removed the [WIP].  No idea what is the problem with the man page.  Ran it locally and it passed.  Just rebased and repushed.  Will see what happens.

**Comment by vrothberg** [comment]
The manpage checks fail because podman errors:
```
[+0000s] checking: podman 
[+0000s] panic: runtime error: invalid memory address or nil pointer dereference
[+0000s] [signal SIGSEGV: segmentation violation code=0x1 addr=0x220 pc=0x1662021]
```

**Comment by vrothberg** [comment]
`./bin/podman --help` triggers it locally as well

File: cmd/podman/commands.go [hunk 1]
```diff
@@ -80,3 +89,107 @@ func getSystemSubCommands() []*cobra.Command {
...
+	}
+	return sysinfo.GetDefaultPidsLimit()
+}
+
+func getDefaultPidsDescription() string {
+	return "Tune container pids limit (set 0 for unlimited)"
+}
+
+func getDefaultDetachKeys() string {
+	return defaultContainerConfig.Libpod.DetachKeys
```
Comment (vrothberg) [inline cmd/podman/commands.go]: The segfault happens here. Looks like `defaultContainerConfig` doesn't exist (yet).

**Comment by rhatdan** [comment]
Just one more bug to get this merged, but I am at a loss to what is causing this. Of course it does not happen locally.
@edsantiago Any ideas?

**Comment by vrothberg** [comment]
Looks like a flake. I restarted to check.

**Comment by rhatdan** [comment]
Well it is a very consistent flake.  Seems to be happening all the time.

**Comment by vrothberg** [comment]
Looking into it. Did you try as remote?

**Comment by vrothberg** [comment]
I can't even test it on my machine. The remote-client doesn't even work properly on master. I see control characters when running a container and build is failing trying to copy data over the varlink bridge.

**Comment by edsantiago** [comment]
```
[+0127s] # error running container: error configuring network list if1 for [/bin/sh -c apk add nginx]: failed to create bridge "do-not-use-google-subnets": could not add "do-not-use-google-subnets": numerical result out of range
```
`contrib/cirrus/99-do-not-use-google-subnets.conf` is copied into `/etc/cni/net.d` for CI. My first guess, and it's just a wild guess, is that something has changed in containernetworking-plugins. I'll look into it further this morning as time allows, but wanted to throw this out in case it makes sense to anyone.

**Comment by rhatdan** [comment]
Well I have attempted to get this to happen on my local box and have tried to manually trigger it but nothing. If we could get a local reproducer, this would be a lot easier to debug.

**Comment by rhatdan** [comment]
Fixes: https://github.com/containers/libpod/issues/3587

**Comment by TomSweeneyRedHat** [comment]
@rhatdan needs a rebase

**Comment by vrothberg** [comment]
Crap. I didn't see that there's a `common/pkg/caps` ... now we have `libpod/pkg/capabilities` and that.

**Comment by vrothberg** [comment]
I'll merge the two.

**Comment by vrothberg** [comment]
https://github.com/containers/common/pull/65

**Comment by rhatdan** [comment]
@TomSweeneyRedHat @mheon @baude @jwhonce @vrothberg @giuseppe PTAL
Finally passes tests.  Not sure what is wrong with rdoproject.org.

**Comment by mheon** [comment]
Please don't merge today, before I cut 1.8.2 final. Will do a review this afternoon.

**Comment by rhatdan** [comment]
/hold
@mheon is in charge of cancelling, unless I get sent back to the drawing board

**Comment by mheon** [comment]
/hold cancel

Full review delayed until tomorrow

**Comment by rhatdan** [comment]
@mheon @vrothberg @giuseppe @TomSweeneyRedHat @QiWang19 @baude Could some give this a review, so we can finally get it merged...

File: cmd/podman/shared/create.go [hunk 2]
```diff
@@ -635,33 +636,46 @@ func ParseCreateOpts(ctx context.Context, c *GenericCLIResults, runtime *libpod.
...
+	if c.Changed("dns-search") {
+		dnsSearches = c.StringSlice("dns-search")
+		// Check for explicit dns-search domain of ''
+		if len(dnsSearches) == 0 {
+			return nil, errors.Errorf("'' is not a valid domain")
+		}
+		// Check for . and dns-search domains
+		if util.StringInSlice(".", dnsSearches) {
+			if len(dnsSearches) > 1 {
+				return nil, errors.Errorf("cannot pass additional search domains when also specifying '.'")
```
Comment (vrothberg) [inline cmd/podman/shared/create.go]: We could move the check down to `if dom == "."` in the following loop. Would avoid iterating twice.

File: docs/source/markdown/podman-run.1.md [hunk 3]
```diff
@@ -867,6 +869,8 @@ Set the user namespace mode for the container.  It defaults to the **PODMAN_USER
 - **host**: run in the user namespace of the caller. This is the default if no user namespace options are set. The processes running in the container will have the same privileges on the host as any other process launched by the calling user.
 - **keep-id**: creates a user namespace where the current rootless user's UID:GID are mapped to the same values in the container. This option is ignored for containers created by the root user.
 - **ns**: run the container in the given existing user namespace.
+- **private**: create a new namespace for the container (default)
+- **private**:
```
Comment (vrothberg) [inline docs/source/markdown/podman-run.1.md]: This line can be deleted.

File: pkg/util/utils.go [hunk 4]
```diff
@@ -609,3 +609,46 @@ func Tmpdir() string {
 
 	return tmpdir
 }
+
+// ValidateSysctls validates a list of sysctl and returns it.
+func ValidateSysctls(strSlice []string) (map[string]string, error) {
```
Comment (vrothberg) [inline pkg/util/utils.go]: Looks like a candidate for a `c/common/pkg/sysctl`.

File: docs/source/markdown/podman-run.1.md (hunk 3 above)
Comment (rhatdan) [reply docs/source/markdown/podman-run.1.md]: Fixed

File: pkg/util/utils.go (hunk 4 above)
Comment (rhatdan) [reply pkg/util/utils.go]: WIll open a different PR for this, but I would prefer to get this merged and fixed later.

Comment (vrothberg) [reply pkg/util/utils.go]: Completely agree :+1:

File: cmd/podman/shared/create.go (hunk 2 above)
Comment (rhatdan) [reply cmd/podman/shared/create.go]: Fixed

File: pkg/util/utils.go (hunk 4 above)
Comment (rhatdan) [reply pkg/util/utils.go]: https://github.com/containers/common/pull/97

File: libpod/config/config.go [hunk 5]
```diff
@@ -1,618 +0,0 @@
...
-				logrus.Warnf("Can not find crun package on the host, containers might fail to run on cgroup V2 systems without crun: %q", err)
-			} else {
-				config.CgroupCheck = true
-				config.OCIRuntime = path
-			}
-		}
-	}
-
-	// If we need to, switch to cgroupfs and logger=file on rootless.
-	config.checkCgroupsAndLogger()
```
Comment (QiWang19) [inline libpod/config/config.go]: should use `func (c *Config) CheckCgroupsAndAdjustConfig()` directly in podman. https://github.com/containers/common/pull/80 removed it from `NewConfig`

Comment (rhatdan) [reply libpod/config/config.go]: Good Catch,
Fixed

**Comment by rhatdan** [comment]
@mheon @vrothberg @QiWang19 @TomSweeneyRedHat @baude @jwhonce PTAL
Fixed up issues.

**Comment by umohnani8** [comment]
LGTM

**Comment by rhatdan** [comment]
@mheon @vrothberg @QiWang19 @TomSweeneyRedHat @baude @jwhonce @giuseppe PTAL
Could we get this in...

**Comment by vrothberg** [comment]
@baude @jwhonce @mheon @rhatdan ... while working on pull I found that the v2 binary is now failing for all invocation with `Error: eventer creation: No support for journald logging`. Looks like this PR introduced it.

I will look into it but we _really_ need some sort of CI for podmanV2.

**Comment by vrothberg** [comment]
K. It turns out we just need to specify the `systemd` build tag.

**Pull request merged**